nomad/state/state_store.go

```diff
@@ -989,26 +997,73 @@ func (s *StateStore) updateNodeDrainImpl(txn *txn, index uint64, nodeID string,
 
 	// Copy the existing node
 	existingNode := existing.(*structs.Node)
-	copyNode := existingNode.Copy()
-	copyNode.StatusUpdatedAt = updatedAt
+	updatedNode := existingNode.Copy()
+	updatedNode.StatusUpdatedAt = updatedAt
 
 	// Add the event if given
 	if event != nil {
-		appendNodeEvents(index, copyNode, []*structs.NodeEvent{event})
+		appendNodeEvents(index, updatedNode, []*structs.NodeEvent{event})
 	}
 
 	// Update the drain in the copy
-	copyNode.DrainStrategy = drain
+	updatedNode.DrainStrategy = drain
 	if drain != nil {
-		copyNode.SchedulingEligibility = structs.NodeSchedulingIneligible
+		updatedNode.SchedulingEligibility = structs.NodeSchedulingIneligible
 	} else if markEligible {
-		copyNode.SchedulingEligibility = structs.NodeSchedulingEligible
+		updatedNode.SchedulingEligibility = structs.NodeSchedulingEligible
+	}
+
+	// Update LastDrain
+	updateTime := time.Unix(updatedAt, 0)
+
+	// if drain strategy isn't set before or after, this wasn't a drain operation
+	// in that case, we don't care about .LastDrain
+	drainNoop := existingNode.DrainStrategy == nil && updatedNode.DrainStrategy == nil
+	// otherwise, when done with this method, updatedNode.LastDrain should be set
+	// if starting a new drain operation, create a new LastDrain. otherwise, update the existing one.
+	startedDraining := existingNode.DrainStrategy == nil && updatedNode.DrainStrategy != nil
+	if !drainNoop {
+		if startedDraining {
+			updatedNode.LastDrain = &structs.DrainMetadata{
+				StartedAt: updateTime,
+				Meta:      drainMeta,
+			}
+		} else if updatedNode.LastDrain == nil {
+			// if already draining and LastDrain doesn't exist, we need to create a new one
+			// this could happen if we upgraded to 1.1.x during a drain
+			updatedNode.LastDrain = &structs.DrainMetadata{
+				// we don't have sub-second accuracy on these fields, so truncate this
+				StartedAt: time.Unix(existingNode.DrainStrategy.StartedAt.Unix(), 0),
```

Comment:
Can't this just be `Started: existingNode.DrainStrategy.StartedAt`? ...not sure why we're converting to a unix timestamp and back.

Reply:
to round down (truncate to seconds). this (may be misremembering) is because updateTime doesn't have the necessary accuracy and there were some funny errors that violated order of events when drain completed immediately. so StartedAt and UpdatedAt are explicitly second-resolution (which means maybe they shouldn't be time.Time 🤔)